witness/src/Dscp/Snowdrop/Expanders.hs

```diff
@@ -306,6 +325,94 @@ seqExpandersBalanceTx feesReceiverAddr (Fees minimalFees) =
+        let issuerAddr = mkAddr (hIssuer header)
+        let isGenesis = header == bHeader genesisBlock
+
+        {- Validation -}
+
+        when isGenesis $
+            -- further we assume that our genesis block has the least possible difficulty
+            unless (hDifficulty header == 0) $
+                error "Genesis block difficulty must be 0"
+
+        unless isGenesis $ do
+            mSimilarHeaderHash <- queryOne (hDifficulty header)
+            case mSimilarHeaderHash of
+                Nothing -> pass
+                Just similarHeaderHash ->
+                    throwLocalError DuplicatedDifficulty
+                    { bmeProvidedHeader = header, bmeExistingHeaderHash = similarHeaderHash }
+
+            -- At this point there is no block in chain with exactly the same
+            -- header as ours
+
+            prevHash <-
+                (hDifficulty header - 1)
+                `assertExists` DifficultyIsTooLarge (hDifficulty header)
+            (sBlockReconstruct -> Block prevHeader _) <-
+                SD.BlockRef prevHash `assertExists` noPrevBlock prevHash
+
+            unless (hPrevHash header == prevHash) $
+                  throwLocalError PrevBlockIsIncorrect
+                  { bmeProvidedHash = hPrevHash header, bmeTipHash = prevHash }
+            unless (hSlotId header > hSlotId prevHeader) $
+                  throwLocalError SlotIdIsNotIncreased
+                  { bmeProvidedSlotId = hSlotId header, bmeTipSlotId = hSlotId prevHeader }
+
+            let GovCommittee com = gcGovernance $ giveL @CoreConfig
+            unless (committeeOwnsSlot com issuerAddr (hSlotId header)) $
+                throwLocalError $ IssuerDoesNotOwnSlot
+                { bmrSlotId = hSlotId header, bmrIssuer = issuerAddr }
+
+        let signedData = BlockToSign (hDifficulty header) (hSlotId header)
+                                     (hPrevHash header) (hash body)
+        unless (verify (hIssuer header) signedData (hSignature header)) $
+            throwLocalError InvalidBlockSignature
+
+        {- Change sets -}
+
+        let diffChange = Map.singleton
+                (inj $ hDifficulty header)
+                (New . BlockIdxVal $ headerHash header)
+
+        -- no keys collision is possible because at this point hPrevHash matches tip,
+        -- and block at tip is unique
+        let forwardBlockChange = Map.singleton
+                (inj . NextBlockOf $ hPrevHash header)
+                (New . NextBlockOfVal . NextBlock $ headerHash header)
+
+        -- no keys collision is possible sinse we have validated transactions
+        -- before this function call, thus every transaction is unique
+        let txBlockRefsChange = Map.fromList $
+                enumerate (bbTxs (bBody block)) <&> \(idx, gTx) ->
+                    ( inj . TxBlockRefId . toGTxId . unGTxWitnessed $ gTx
+                    , New . TxBlockVal $ TxBlockRef (headerHash block) idx
+                    )
```

Comment:
Can we collect those changes in transactions expanders? E. g. by passing header hash to `expandGTxs` and further to `seqExpandersGTx`? I'm uncomfortable with the fact that this is the only reason why `BlockMetaTx` contains the whole `Block` and not only its header is this piece of code.

Reply:
Good point :+1:

Reply:
Hmm, actually also I moved signature verification to the expander, and block body participates in it.
I also find carrying the whole block around ugly, but ways to make it less ugly doesn't seem to be worth the effort :thinking:

Reply:
@Martoon-00 Just found out that the block body is hashed _entirely_ instead of building a Merkle tree over it...
WTF is really going on, my god, I really need to dive into our public chain code, the more I look into it, the more ashamed and confused I become.

Reply:
And actually, there shouldn't be even a Merkle tree hash, there should be just a _final AVL root_ in the block header which should be signed, and this will already give us all the guarantees. Where is this final AVL hash stored in the block anyway? I don't see any.

Reply:
@Martoon-00 nevermind, this totally should not be addressed in scope of this PR.

Reply:
Nice, I'll set a todo to keep just the header in `TxBlockMeta`

Reply:
in the corresponding YT ticket